### ISSUE TITLE
Work around linker issues in ghcup version of GHC 7.10.3

### ISSUE THREAD

**Issue by RyanGlScott** [haskell-CI/haskell-ci#543]
As reported in https://gitlab.haskell.org/haskell/ghcup-hs/-/issues/123, the version of GHC 7.10.3 that `ghcup` provides is susceptible to linker errors whenever it compiles any executable. ([Here](https://github.com/RyanGlScott/gha-sandbox/runs/3719228842?check_suite_focus=true#step:19:32) are examples of such linker errors.) My guess is that the `hvr-ppa` version of GHC 7.10.3 was patched to avoid this issue, and `ghcup` recommends [this workaround](https://gitlab.haskell.org/haskell/ghcup-hs/-/commit/ec7130dac6ed75c54320fe7eef37e981937368c2) after installing 7.10.3. We should apply such a workaround in `haskell-ci`. Something like this:

```diff
diff --git a/src/HaskellCI/GitHub.hs b/src/HaskellCI/GitHub.hs
index a7767e2..cc8e4bb 100644
--- a/src/HaskellCI/GitHub.hs
+++ b/src/HaskellCI/GitHub.hs
@@ -261,6 +261,14 @@ makeGitHub _argv config@Config {..} gitconfig prj jobs@JobVersions {..} = do
                 , "  url: http://hackage.haskell.org/"
                 ]

+            -- Using -optl-no-pie is required to work around
+            -- https://gitlab.haskell.org/haskell/ghcup-hs/-/issues/123
+            when ((GHC (C.mkVersion [7,10,3]) `elem` allVersions) &&
+                  (GHC (C.mkVersion [7,10,3]) & isGHCUP)) $ do
+                let range = Range (C.thisVersion (C.mkVersion [7,10,3])) /\ Range cfgGhcupJobs
+                echo_if_to range "$CABAL_CONFIG" "program-default-options"
+                echo_if_to range "$CABAL_CONFIG" "  ghc-options: -optl-no-pie"
+
             -- Add head.hackage repository to ~/.cabal/config
             -- (locally you want to add it to cabal.project)
             unless (S.null headGhcVers) $ sh $ concat $
```

**Comment by phadej** [comment]
I'm tempted to just not recommend using `ghcup` for old(er) GHCs. They are available from hvr-ppa. Also simply not supporting ghc 7.10 is IMHO a viable option. (or if you want to support older, then `ghcup` is not going any further back...)

**Comment by phadej** [comment]
I pressed comment too early.

In general, I would like to avoid this general class of problems, by simply doing nothing. Official GHC bindists "bitrot", in a sense, that they aren't forward compatible with toolchain, os, etc changes. This issue AFAIK happened because newer debians and ubuntus changed how linker works by default, and hvr-ppa was the right place to fix that, in packaging for these distributions.

(I.e. that flag is not necessary if you use old debian/ubuntu distribution!)

TL;DR, I don't want to do "packaging" in `haskell-ci`. IMO `ghcup` could just refuse to install GHC-7.10 on newer debian derivatives.

**Comment by RyanGlScott** [comment]
> I don't want to do "packaging" in `haskell-ci`.

We're already doing packaging: see [these lines](https://github.com/haskell-CI/haskell-ci/blob/2175f7b1cf4cf2f0c0d7ef6647cbc439755c4f3e/src/HaskellCI/GitHub.hs#L140-L145), which are essential to using GHC 8.4.4 with GHCup.

To be clear, I would favor dropping support for GHC versions if it becomes unfeasible to work around issues like these. But currently, the steps needed to install GHC 7.10.3 aren't so bad, and certainly not any worse than the existing steps we already use to install GHC 8.4.4.

**Comment by phadej** [comment]
> We're already doing packaging: see these lines, which are essential to using GHC 8.4.4 with GHCup.

Yes, and I suspect it will become worse and worse over the time. This is why I'm not thrilled.

**Comment by sjakobi** [comment]
Please note that HVR's PPA _doesn't_ provide GHC 7.10.3 for Ubuntu Bionic: https://launchpad.net/~hvr/+archive/ubuntu/ghc?field.series_filter=bionic

I think this is the reason why I'm seeing [this issue](https://github.com/quchen/prettyprinter/pull/208/checks?check_run_id=3743288106#step:6:73) even though my GHA config shows `setup-method: hvr-ppa`.

I'd very much like to get a solution in `haskell-ci` – dropping CI with GHC 7.10.3 is not a good option for `prettyprinter` yet.

**Comment by phadej** [comment]
Don't it?
![Screenshot from 2021-09-29 14-36-30](https://user-images.githubusercontent.com/51087/135261038-917da0fb-d727-4e7d-90dc-5a9e66aec2bb.png)

> I think this is the reason why I'm seeing this issue even though my GHA config shows setup-method: hvr-ppa.

it's because  `haskell/actions/setup` uses `ghcup` to install GHc

![Screenshot from 2021-09-29 14-39-25](https://user-images.githubusercontent.com/51087/135261289-e5f1124e-172e-43c6-832f-969a5603baf0.png)

i.e. you @sjakobi should use the same workaround in your CI script as Ryan links to in the initial comment.

**Comment by sjakobi** [comment]
Thanks, @phadej. I should have looked harder! :facepalm: 

I can try that patch. My preference would be to install the patched GHC from `hvr-ppa` though. How do I do that?

**Comment by sjakobi** [comment]
> I can try that patch. My preference would be to install the patched GHC from `hvr-ppa` though. How do I do that?

Please disregard. I just noticed that the failing job wasn't configured via `haskell-ci`. :facepalm: again. :/
